### PR TITLE
Refactor transformation Blocks

### PR DESCRIPTION
**Pull request by webrouse**
Part of: https://keboola.atlassian.net/browse/COM-1138

Prvy commit, implementacia, ~300 riadkov.
Druhy commit, upravene/doplnene testy.

### CODE REVIEW

File: internal/pkg/model/object.go [hunk 1]
```diff
@@ -110,7 +114,7 @@ type Config struct {
 	Description       string                 `json:"description" diff:"true" descriptionFile:"true"`
 	ChangeDescription string                 `json:"changeDescription"`
 	Content           *orderedmap.OrderedMap `json:"configuration" validate:"required" diff:"true" configFile:"true"`
-	Blocks            Blocks                 `json:"-" validate:"dive"` // loaded transformation's blocks, filled in only for the LOCAL state
```
Comment (webrouse) [inline internal/pkg/model/object.go:113]: `Config.Blocks` su taky kostlivec, ktory nejak fungoval, ale nebol spraveny poriadne.
```
 // loaded transformation's blocks, filled in only for the LOCAL state
```

Transformation blocks su na disku reprezentovane ako adresar `blocks`, v API su sucastou konfiguracie, pod klucom `blocks`:
https://github.com/keboola/keboola-as-code/tree/main/test/pull/out/bar/transformation/keboola.python-transformation-v2/python-transformation

**Fungovalo to tak:**
- pri local load 
  - sa nacitali `blocks` z adresara a ulozili sa priamo do `config.Content` pod kluc `blocks` ako `orderedMap`.
  - zaroven sa ulozili do `Config.Blocks` ako `[]*Block`, aby sme pri `rename` vedeli, aka blocks lokalne existuju.
  - tj `local state -> api state`
- pri remote load
  - sa nedialo nic
  - diff `blocks` sa robil vramci `Config.Content`

**Upravil som to:**
- local load nacita adresar `blocks` a ulozi ho do `Config.Blocks` (internal representation)
- remote load nacita `blocks` z `Config.Content` a ulozi do `Config.Blocks` (internal representation)
- diff sa robi nad `Config.Blocks`
- a save remote/local funguje ako load len opacne

File: internal/pkg/model/object.go [hunk 2]
```diff
@@ -154,20 +158,20 @@ type Blocks []*Block
 // Block - transformation block.
 type Block struct {
 	BlockKey
-	Paths `json:"-"`
-	Name  string `json:"name" validate:"required" metaFile:"true"`
-	Codes Codes  `json:"codes" validate:"omitempty,dive"`
+	PathInProject `json:"-"`
```
Comment (webrouse) [inline internal/pkg/model/object.go:161]: Block nema ziadne `RelatedPaths`, preto nahradene s `PathInProject`.

File: internal/pkg/model/object.go [hunk 3]
```diff
@@ -154,20 +158,20 @@ type Blocks []*Block
...
 type Codes []*Code
 
 // Code - transformation code.
 type Code struct {
 	CodeKey
-	Paths        `json:"-"`
-	CodeFileName string   `json:"-"` // eg. "code.sql", "code.py", ...
-	Name         string   `json:"name" validate:"required" metaFile:"true"`
-	Scripts      []string `json:"script"` // scripts, eg. SQL statements
+	PathInProject `json:"-"`
```
Comment (webrouse) [inline internal/pkg/model/object.go:171]: Code nema ziadne RelatedPaths, preto nahradene s PathInProject.

File: internal/pkg/model/object.go [hunk 4]
```diff
@@ -369,3 +373,29 @@ func (c *Config) AddRelation(relation Relation) {
 func (r *ConfigRow) AddRelation(relation Relation) {
 	r.Relations.Add(relation)
 }
+
+func (b Block) String() string {
```
Comment (webrouse) [inline internal/pkg/model/object.go:377]: Block/Code -> `string`, pouziva sa v diffe.

File: internal/pkg/remote/manager.go [hunk 5]
```diff
@@ -155,7 +148,14 @@ func (u *UnitOfWork) loadObject(object model.Object) (model.ObjectState, error)
 	}
 
 	// Set remote state
-	objectState.SetRemoteState(object)
+	internalObject := object.Clone()
+	objectState.SetRemoteState(internalObject)
+
+	// Invoke mapper
+	recipe := &model.RemoteLoadRecipe{Manifest: objectState.Manifest(), ApiObject: object, InternalObject: internalObject}
```
Comment (webrouse) [inline internal/pkg/remote/manager.go:155]: Do `RemoteLoadRecipe` pridany `Manifest`, aby sme vedeli generovat `path` ku Block/Code.
Preto `Invoke mapper` presunute nizsie, az ked mame k dispozicii `objectState.Manifest()`.

File: internal/pkg/utils/map.go [hunk 6]
```diff
@@ -62,6 +62,30 @@ func CloneOrderedMap(in *orderedmap.OrderedMap) *orderedmap.OrderedMap {
 	return out
 }
 
+func GetFromMap(m *orderedmap.OrderedMap, keys []string) interface{} {
```
Comment (webrouse) [inline internal/pkg/utils/map.go:65]: Utilitka, ktora nacita z `orderedMap` nejaku nested hodnotu, ... napr `foo.bar -> []string{foo, bar}`, vid testy.

File: internal/pkg/utils/map_test.go [hunk 7]
```diff
@@ -84,3 +87,41 @@ func TestUpdateSliceStep(t *testing.T) {
...
+	}
+}
+`
+	m := NewOrderedMap()
+	json.MustDecodeString(input, m)
+	cases := []struct {
+		keys     []string
+		expected interface{}
+	}{
+		{[]string{}, `{"foo":{"bar":{"baz":{"value":123}}}}`},
```
Comment (webrouse) [inline internal/pkg/utils/map_test.go:110]: Testy pre `GetFromMap`.

File: internal/pkg/diff/diff.go [hunk 8]
```diff
@@ -172,6 +172,10 @@ func (d *Differ) newOptions(reporter *Reporter) cmp.Options {
 		cmpopts.AcyclicTransformer("relations", func(relations model.Relations) model.RelationsBySide {
 			return relations.RelationsBySide()
 		}),
+		// Dif transformation blocks as string
+		cmp.Transformer("block", func(block model.Block) string {
```
Comment (webrouse) [inline internal/pkg/diff/diff.go:176]: Block sa porovnava ako stirng, vid test.

File: internal/pkg/diff/diff_test.go [hunk 9]
```diff
@@ -418,6 +418,159 @@ func TestDiffRelations(t *testing.T) {
...
+				},
+				PathInProject: model.NewPathInProject(`branch/config/blocks`, `001-my-block`),
+			},
+		},
+	}
+	objectState.SetLocalState(lObject)
+
+	differ := NewDiffer(projectState)
+	reporter := differ.diffValues(objectState, rObject.Blocks, lObject.Blocks)
+	expected := `
```
Comment (webrouse) [inline internal/pkg/diff/diff_test.go:490]: Diff `blocks`, vysledok aj predchadzajuceho PR.

File: internal/pkg/diff/diff_test.go [hunk 10]
```diff
@@ -418,6 +418,159 @@ func TestDiffRelations(t *testing.T) {
...
+  blocks/001-my-block:
```
Comment (webrouse) [inline internal/pkg/diff/diff_test.go:491]: Namiesto indexu, radsej vypiseme uzivatelovi testu k adresaru.

File: internal/pkg/diff/diff_test.go [hunk 11]
```diff
@@ -418,6 +418,159 @@ func TestDiffRelations(t *testing.T) {
...
+    - #  Block 1
+    + #  My block
+      ## Code 1
+      SELECT 1;
+    + SELECT 2;
+    + SELECT 3;
```
Comment (webrouse) [inline internal/pkg/diff/diff_test.go:497]: Block 1 sa zmenil.

File: internal/pkg/diff/diff_test.go [hunk 12]
```diff
@@ -418,6 +418,159 @@ func TestDiffRelations(t *testing.T) {
...
+- blocks/001-block-1/001-code-1:
+-   #  Block 2
+-   ## Code 2
+-   SELECT 2;
+-
```
Comment (webrouse) [inline internal/pkg/diff/diff_test.go:502]: Block2 je iba v remote state.

File: internal/pkg/diff/diff_test.go [hunk 13]
```diff
@@ -418,6 +418,159 @@ func TestDiffRelations(t *testing.T) {
...
+				}),
+			},
+		}),
+	}
+	objectState.SetLocalState(lObject)
+
+	differ := NewDiffer(projectState)
+	reporter := differ.diffValues(objectState, rObject.Content, lObject.Content)
+	expected := `
+  foo.bar:
```
Comment (webrouse) [inline internal/pkg/diff/diff_test.go:561]: Ak v nejakej map boli rozne typy hodnot, tak sa negeroval spravny diff -> pridany test.
Porovnanie `string ` a `map`.

File: internal/pkg/diff/reporter.go [hunk 14]
```diff
@@ -200,32 +201,47 @@ func (r *Reporter) objectPath(value reflect.Value) string {
...
 
+func formatValue(value reflect.Value) []string {
+	if value.Type().Kind().String() == `interface` {
+		value = value.Elem()
+	}
+
+	var formatted string
+	if strings.HasPrefix(value.Type().String(), "map[") {
+		// Format map to JSON
+		formatted = strings.TrimRight(json.MustEncodeString(value.Interface(), true), "\n")
```
Comment (webrouse) [inline internal/pkg/diff/reporter.go:238]: Chybalo formatiovanie map.

File: internal/pkg/mapper/transformation/load.go [hunk 15]
```diff
@@ -13,16 +13,80 @@ import (
...
 
-type loader struct {
-	model.MapperContext
-	*model.LocalLoadRecipe
-	config    *model.Config
-	blocksDir string
-	blocks    []*model.Block
-	errors    *utils.Error
+// MapAfterRemoteLoad - load code blocks from API to blocks field.
+func (m *transformationMapper) MapAfterRemoteLoad(recipe *model.RemoteLoadRecipe) error {
```
Comment (webrouse) [inline internal/pkg/mapper/transformation/load.go:17]: Mapovanie `Config.Content.blocks` -> `Config.Blocks` pri remote load.

File: internal/pkg/mapper/transformation/load.go [hunk 16]
```diff
@@ -13,16 +13,80 @@ import (
...
+			code.BranchId = config.BranchId
+			code.ComponentId = config.ComponentId
+			code.ConfigId = config.Id
+			code.BlockIndex = block.Index
+			code.Index = codeIndex
+		}
+	}
+
+	// Set paths, if are present
+	if recipe.Manifest.Path() != "" {
```
Comment (webrouse) [inline internal/pkg/mapper/transformation/load.go:67]: Ak sa pouziva existujuci manifest, tak ma path -> mozu existovat aj nejake blocks.

File: internal/pkg/mapper/transformation/load.go [hunk 17]
```diff
@@ -13,16 +13,80 @@ import (
...
+		blocksDir := m.Naming.BlocksDir(recipe.Manifest.Path())
+		for _, block := range config.Blocks {
+			if path, found := m.Naming.GetCurrentPath(block.Key()); found {
```
Comment (webrouse) [inline internal/pkg/mapper/transformation/load.go:70]: Ak block uz existuje, pouzijeme jeho cestu, ... inak vygenerujeme novu.

File: internal/pkg/mapper/transformation/save.go [hunk 18]
```diff
@@ -2,26 +2,64 @@ package transformation
...
 )
 
-type writer struct {
-	model.MapperContext
-	*model.LocalSaveRecipe
-	config    *model.Config
-	configDir string
-	errors    *utils.Error
+// MapBeforeRemoteSave - save code blocks to the API.
+func (m *transformationMapper) MapBeforeRemoteSave(recipe *model.RemoteSaveRecipe) error {
```
Comment (webrouse) [inline internal/pkg/mapper/transformation/save.go:15]: Mapovanie `Config.Blocks` -> `Config.Content.blocks` pri remote save.

File: internal/pkg/mapper/transformation/save.go [hunk 19]
```diff
@@ -138,16 +150,3 @@ func (w *writer) createMetadataFile(path, desc string, content *orderedmap.Order
 		w.errors.Append(err)
 	}
 }
-
-func (w *writer) joinScripts(scripts []string) string {
```
Comment (webrouse) [inline internal/pkg/mapper/transformation/save.go:142]: Presunute do `Code.ScriptsToString`.

File: internal/pkg/mapper/transformation/save_test.go [hunk 20]
```diff
@@ -1,19 +1,122 @@
...
 
 	"github.com/keboola/keboola-as-code/internal/pkg/filesystem"
 	"github.com/keboola/keboola-as-code/internal/pkg/json"
 	. "github.com/keboola/keboola-as-code/internal/pkg/mapper/transformation"
+	"github.com/keboola/keboola-as-code/internal/pkg/model"
 	"github.com/keboola/keboola-as-code/internal/pkg/utils"
 )
 
-func TestSaveTransformationEmpty(t *testing.T) {
+func TestRemoteSaveTransformation(t *testing.T) {
```
Comment (webrouse) [inline internal/pkg/mapper/transformation/save_test.go:16]: Test pre `remote save`.

File: internal/pkg/mapper/transformation/save_test.go [hunk 21]
```diff
@@ -26,54 +129,107 @@ func TestSaveTransformationEmpty(t *testing.T) {
 	assert.Equal(t, `{}`, configContent)
 }
 
-func TestSaveTransformation(t *testing.T) {
+func TestLocalSaveTransformation(t *testing.T) {
```
Comment (webrouse) [inline internal/pkg/mapper/transformation/save_test.go:132]: Upraveny test pre `local save`.

File: internal/pkg/mapper/transformation/load_test.go [hunk 22]
```diff
@@ -54,9 +54,9 @@ func TestLoadTransformationMissingCodeMeta(t *testing.T) {
 	}, "\n"), err.Error())
 }
 
-func TestLoadTransformationSql(t *testing.T) {
+func TestLoadLocalTransformationSql(t *testing.T) {
```
Comment (webrouse) [inline internal/pkg/mapper/transformation/load_test.go:57]: Upraveny test pre `local load`.

File: internal/pkg/mapper/transformation/load_test.go [hunk 23]
```diff
@@ -171,48 +228,252 @@ func TestLoadTransformationPy(t *testing.T) {
...
+				`003-block-3`,
+			),
+			Name:  "003",
+			Codes: model.Codes{},
+		},
+	}
+	assert.Equal(t, expected, config.Blocks)
+}
+
+func TestLoadRemoteTransformation(t *testing.T) {
```
Comment (webrouse) [inline internal/pkg/mapper/transformation/load_test.go:334]: Novy test pre `remote load` (odporucam side by side diff).

File: test/pull-changed/expected-stdout [hunk 24]
```diff
@@ -1,6 +1,6 @@
 Plan for "pull" operation:
   * C bar/extractor/ex-generic-v2/empty | changed: configuration
   * R bar/extractor/keboola.ex-db-mysql/with-rows/rows/disabled | changed: configuration, isDisabled
-  * C bar/transformation/keboola.python-transformation-v2/python-transformation | changed: configuration, description
+  * C bar/transformation/keboola.python-transformation-v2/python-transformation | changed: blocks, description
```
Comment (webrouse) [inline test/pull-changed/expected-stdout:4]: Vo funkcionalnych testoch, iba tato drobna zmena, inak sa to sprava rovnako.
Ale `Config.Blocks` su definovane pre remote aj local stave -> daju sa na tom stavat dalsie veci.

**Review by JakubMatejka — APPROVED** [review]

**Pull request merged**